### PR TITLE
Clarify call-by-value

### PR DESCRIPTION
**Pull request by spazmodius**
#### Summary

The wording in Javascript/Guide/Functions confusingly implies that only primitives are passed by value, then apparently contrasts this with changing a property of an object via a passed reference.

Fix the wording so these are clearly orthogonal concerns.

#### Motivation

MDN has been [cited](https://news.ycombinator.com/item?id=16234789) for getting Javascript parameter passing wrong.

#### Metadata

- [ ] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error
- [X] Rewords a misleading explanation

### CODE REVIEW

File: files/en-us/web/javascript/guide/functions/index.md [hunk 1]
```diff
@@ -38,9 +38,9 @@ The function `square` takes one parameter, called `number`. The function consist
 return number * number;
 ```
 
-Primitive parameters (such as a number) are passed to functions **by value**; the value is passed to the function, but if the function changes the value of the parameter, **this change is not reflected globally or in the calling function**.
+Parameters (such as a number or an object reference) are passed to functions **by value**; the value is passed to the function, but if the function changes the value of the parameter, **this change is not reflected globally or in the calling function**.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/guide/functions/index.md:41]:
```suggestion
Parameters are passed to functions **by value**; the value is passed to the function, but if the function then changes the value of the parameter, **the change is not reflected globally or in the caller**.
```

File: files/en-us/web/javascript/guide/functions/index.md [hunk 2]
```diff
@@ -38,9 +38,9 @@ The function `square` takes one parameter, called `number`. The function consist
...
 
-If you pass an object (i.e., a non-primitive value, such as {{jsxref("Array")}} or a user-defined object) as a parameter and the function changes the object's properties, that change is visible outside the function, as shown in the following example:
+If you pass a reference (i.e., to a non-primitive value, such as {{jsxref("Array")}} or a user-defined object) as a parameter and the function changes the object's properties, that change is visible outside the function, as shown in the following example:
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/guide/functions/index.md:43]:
```suggestion
If as a parameter you pass a reference — which will necessarily be a reference to an object of some kind — and the function changes the object's properties, that change is visible outside the function, as shown in the following example:
```

**Review by sideshowbarker — APPROVED** [review]

File: files/en-us/web/javascript/guide/functions/index.md [hunk 3]
```diff
@@ -38,9 +38,9 @@ The function `square` takes one parameter, called `number`. The function consist
 return number * number;
 ```
 
-Primitive parameters (such as a number) are passed to functions **by value**; the value is passed to the function, but if the function changes the value of the parameter, **this change is not reflected globally or in the calling function**.
+Parameters are passed to functions **by value**; the value is passed to the function, but if the function then changes the value of the parameter, **the change is not reflected globally or in the caller**.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/guide/functions/index.md:41]:
```suggestion
Parameters are passed to functions **by value**; the value is passed to the function, but if the function then changes the value of the parameter, **the change is not reflected globally or in the calling code**.
```

File: files/en-us/web/javascript/guide/functions/index.md [hunk 4]
```diff
@@ -38,9 +38,9 @@ The function `square` takes one parameter, called `number`. The function consist
 return number * number;
 ```
 
-Primitive parameters (such as a number) are passed to functions **by value**; the value is passed to the function, but if the function changes the value of the parameter, **this change is not reflected globally or in the calling function**.
+Parameters are passed to functions **by value**; the value is passed to the function, but if the function then changes the value of the parameter, **the change is not reflected globally or in the calling code**.
 
-If you pass an object (i.e., a non-primitive value, such as {{jsxref("Array")}} or a user-defined object) as a parameter and the function changes the object's properties, that change is visible outside the function, as shown in the following example:
+If as a parameter you pass a reference — which will necessarily be a reference to an object of some kind — and the function changes the object's properties, that change is visible outside the function, as shown in the following example:
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/guide/functions/index.md:43]:
```suggestion
However, when you pass an object as a parameter, if the function changes the object's properties, that change is visible outside the function, as shown in the following example:
```

File: files/en-us/web/javascript/guide/functions/index.md [hunk 5]
```diff
@@ -38,9 +38,9 @@ The function `square` takes one parameter, called `number`. The function consist
 return number * number;
 ```
 
-Primitive parameters (such as a number) are passed to functions **by value**; the value is passed to the function, but if the function changes the value of the parameter, **this change is not reflected globally or in the calling function**.
+Parameters are passed to functions **by value**; the value is passed to the function, but if the function then changes the value of the parameter, **the change is not reflected globally or in the calling code**.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/guide/functions/index.md:41]:
```suggestion
Parameters are passed to functions **by value** — so if the code within the body of a function assigns a completely new value to a parameter that was passed to the function, **the change is not reflected globally or in the code which called that function**.
```

**Review by sideshowbarker — APPROVED** [review]

File: files/en-us/web/javascript/guide/functions/index.md [hunk 6]
```diff
@@ -38,9 +38,9 @@ The function `square` takes one parameter, called `number`. The function consist
 return number * number;
 ```
 
-Primitive parameters (such as a number) are passed to functions **by value**; the value is passed to the function, but if the function changes the value of the parameter, **this change is not reflected globally or in the calling function**.
+Parameters are passed to functions **by value** — so if the code within the body of a function assigns a completely new value to a parameter that was passed to the function, **the change is not reflected globally or in the code which called that function**.
 
-If you pass an object (i.e., a non-primitive value, such as {{jsxref("Array")}} or a user-defined object) as a parameter and the function changes the object's properties, that change is visible outside the function, as shown in the following example:
+However, when you pass an object as a parameter, if the function changes the object's properties, that change is visible outside the function, as shown in the following example:
```
Comment (spazmodius) [inline files/en-us/web/javascript/guide/functions/index.md:43]: Nope, "However" is spurious juxtaposition. Both statements are true, full stop; the second statement does not refine the first.

Comment (sideshowbarker) [reply files/en-us/web/javascript/guide/functions/index.md:43]: > Nope, "However" is spurious juxtaposition. Both statements are true, full stop; the second statement does not refine the first.

OK, thanks — I’ve now fixed that over in https://github.com/mdn/content/pull/12344

**Pull request merged**